dbms/src/Functions/array/arrayReduceInRanges.cpp

```diff
@@ -0,0 +1,402 @@
+                agg_func.destroy(places[j]);
+        });
+
+        auto true_func = &agg_func;
+        /// Unnest consecutive trailing -State combinators
+        while (auto func = typeid_cast<AggregateFunctionState *>(true_func))
+            true_func = func->getNestedFunction().get();
+
+        /// Pre-aggregate to the initial level
+        for (size_t j = 0; j < place_offsets[0]; ++j)
+        {
+            size_t local_begin = j * minimum_step;
+            size_t local_end = (j + 1) * minimum_step;
+
+            for (size_t k = local_begin; k < local_end; ++k)
+                true_func->add(places[j], aggregate_arguments, begin + k, arena.get());
+        }
+
+        /// Pre-aggregate to the higher levels by merging
+        {
+            size_t place_in_level = place_offsets[0] >> 1;
+            size_t place_begin = 0;
+
+            for (size_t level = 0; place_in_level; ++level)
+            {
+                size_t next_place_begin = place_offsets[level];
+
+                for (size_t j = 0; j < place_in_level; ++j)
+                {
+                    true_func->merge(places[next_place_begin + j], places[place_begin + (j << 1)], arena.get());
+                    true_func->merge(places[next_place_begin + j], places[place_begin + (j << 1) + 1], arena.get());
+                }
+
+                place_in_level >>= 1;
+                place_begin = next_place_begin;
+            }
+        }
+
+        for (size_t j = ranges_begin; j < ranges_end; ++j)
+        {
+            size_t local_begin = 0;
+            size_t local_end = 0;
+
+            {
+                Int64 index = indices_col.getInt(j);
+                UInt64 length = lengths_col.getUInt(j);
+
+                /// Keep the same as in arraySlice
+
+                if (index > 0)
+                {
+                    local_begin = index - 1;
+                    if (local_begin + length < end - begin)
+                        local_end = local_begin + length;
+                    else
+                        local_end = end - begin;
+                }
+                else if (index < 0)
+                {
+                    if (end - begin + index > 0)
+                        local_begin = end - begin + index;
+                    else
+                        local_begin = 0;
+
+                    if (local_begin + length < end - begin)
+                        local_end = local_begin + length;
+                    else
+                        local_end = end - begin;
+                }
+            }
+
+            size_t place_begin = (local_begin + minimum_step - 1) / minimum_step;
+            size_t place_end = local_end / minimum_step;
+
+            AggregateDataPtr place = arena->alignedAlloc(agg_func.sizeOfData(), agg_func.alignOfData());
+            try
+            {
+                agg_func.create(place);
+            }
+            catch (...)
+            {
+                agg_func.destroy(place);
```

Comment:
This is wrong. If constructor has thrown an exception, destructor must not be called.